flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/ATSU/A320_Neo_CDU_ATC_OceanicReq.js

```diff
@@ -0,0 +1,331 @@
+class CDUAtcOceanicReq {
+    static CreateDataBlock() {
+        return {
+            entryPoint: null,
+            entryTime: null,
+            requestedMach: null,
+            requestedFlightlevel: null,
+            freetext: [ '', '', '', '', '', '' ]
+        };
+    }
+
+    static CanSendData(mcdu, data) {
+        if (!SimVar.GetSimVarValue("ATC FLIGHT NUMBER", "string", "FMC") || SimVar.GetSimVarValue("ATC FLIGHT NUMBER", "string", "FMC").length === 0) {
+            return false;
+        }
+        if (!mcdu.flightPlanManager.getDestination() || mcdu.flightPlanManager.getDestination().ident === "") {
+            return false;
+        }
+        if (mcdu.atsuManager.atc.currentStation() === "") {
+            return false;
+        }
+        return data.entryPoint && data.entryTime && data.requestedMach && data.requestedFlightlevel;
+    }
+
+    static CreateMessage(mcdu, data) {
+        const retval = new Atsu.OclMessage();
+
+        retval.Callsign = SimVar.GetSimVarValue("ATC FLIGHT NUMBER", "string", "FMC");
+        retval.Destination = mcdu.flightPlanManager.getDestination().ident;
+        retval.EntryPoint = data.entryPoint;
+        retval.EntryTime = data.entryTime;
+        retval.RequestedMach = data.requestedMach;
+        retval.RequestedFlightlevel = data.requestedFlightlevel;
+        retval.Freetext = data.freetext.filter((n) => n);
+        retval.Station = mcdu.atsuManager.atc.currentStation();
+
+        return retval;
+    }
+
+    static WaypointOnRoute(mcdu, ident) {
+        const totalWaypointsCount = mcdu.flightPlanManager.getWaypointsCount() + mcdu.flightPlanManager.getArrivalWaypointsCount() + mcdu.flightPlanManager.getApproachWaypoints().length;
+        const wptsListIndex = mcdu.flightPlanManager.getActiveWaypointIndex();
+        let i = 0;
+
+        while (i < totalWaypointsCount && i + wptsListIndex < totalWaypointsCount) {
+            const waypoint = mcdu.flightPlanManager.getWaypoint(i + wptsListIndex, NaN, true);
+
+            if (waypoint && !waypoint.isVectors && waypoint.ident === ident) {
+                return true;
+            }
+
+            i++;
+        }
+
+        return false;
+    }
+
+    static CalculateEntryPointETA(mcdu, ident) {
+        const adirLat = ADIRS.getLatitude();
+        const adirLong = ADIRS.getLongitude();
+        const ppos = (adirLat.isNormalOperation() && adirLong.isNormalOperation()) ? {
+            lat: ADIRS.getLatitude().value,
+            long: ADIRS.getLongitude().value,
+        } : {
+            lat: NaN,
```

Comment:
Personally not a fan of NaN for this... it's meant to be for floating point errors. Missing data should be null or undefined (I know there are other examples of NaN being misused in many places).

Reply:
fixed